### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by SaFrMo**
Things to test:
- Add a product on the [test Shopify site](https://sander-test.myshopify.com/admin) (login is in 1Password)
- Refresh products on the [WP backend](http://sander.funkhaus.us/wp-admin/tools.php?page=wshop_settings)
- Go to [Products](http://sander.funkhaus.us/wp-admin/edit.php?post_type=wps-product) to view updated products
- Add and remove products on Shopify, then try out the options on the Refresh Products section

### CODE REVIEW

**Comment by johndigital** [comment]
Good work @SaFrMo, all the tests seem to be working for me. A few high level notes:
- I think we only need the "Publish new products" checkbox. The plugin should automatically delete any products that don't exist in Shopify.
- For the console area under the "refresh products" button, let's have new messages come in at the top rather than the bottom. That way the user doesn't have to scroll to see the progress if there is a high number of products.
- Let's change "Permalink Structure Slug:" to be something more readable like "store permalink"
- The whole "Refresh Products" area at the bottom should only show up if all the API info is hooked up correctly (The top 3 settings fields should be filled in)
- Let's move the settings page out of "Tools" and into "Settings." This thing is starting to be a full blown plugin!

Going to have a look through the code now and will add comments inline.

File: js/wshop.js [hunk 1]
```diff
@@ -12,7 +12,8 @@ var wshop = {
 
         wshop.shopClient = ShopifyBuy.buildClient({
             apiKey: wshopVars.apiKey,
-            domain: wshopVars.domain,
+            // Strips out 'http' if user entered it in their options
+            domain: wshopVars.domain.replace(/^https?:\/\//, ''),
```
Comment (johndigital) [inline js/wshop.js]: Let's do this as the URL is being saved to the options table. No sense in running it on every page load!

File: js/wshop.refresh.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        wshopRefresh.initListener();
+    },
+
+    initListener: function(){
+
+        jQuery('#wpshopify-refresh-button').on('click', function(e){
+            e.preventDefault();
+
+            // Cancel if already working
+            if( jQuery('#wpshopify-refresh-button').hasClass('disabled') ) return;
```
Comment (johndigital) [inline js/wshop.refresh.js]: You shouldn't need a 'disabled' class, you should be able to do any styling etc. with the disabled attribute.

File: js/wshop.refresh.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        jQuery.post({
+            url: wshopRefresh.vars.processLink,
+            data: {
+                product_id: data.id,
+                product_title: data.title,
+                auto_publish: jQuery('#auto_approve').is(':checked')
+            }
+        }).done(function(message){
+
+            jQuery('.refresh-message').append(`<li>(${wshopRefresh.totalProducts - wshopRefresh.products.length} / ${wshopRefresh.totalProducts}) ${message}</li>`);
```
Comment (johndigital) [inline js/wshop.refresh.js]: I don't think template literals are supported enough for us to use yet

File: js/wshop.refresh.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            // Strips out 'http' if user entered it in their options
+            domain: wshopVars.domain.replace(/^https?:\/\//, ''),
+            appId: wshopVars.appId
+        });
+    },
+
+    processAllProducts: function(products){
+
+        // Sort Products by ID number
+        products = products.sort(function(a, b){
```
Comment (johndigital) [inline js/wshop.refresh.js]: This seems unnecessary

File: wp-shopify.php [hunk 5]
```diff
@@ -26,4 +26,132 @@ function pp() {
...
+        }
+    }
+
+    register_activation_hook( __FILE__, 'set_wps_rewrite_slug' );
+
+    // Define AJAX functions
+    function wps_process_product(){
+
+        // Get the ID of the current Product
+        $id = $_REQUEST['product_id'];
```
Comment (johndigital) [inline wp-shopify.php]: use $_POST instead of request. We don't want this to be able to happen with a GET request

File: wp-shopify.php [hunk 6]
```diff
@@ -26,4 +26,132 @@ function pp() {
...
+
+        $output = 'Error processing Product ' . $title . '!';
+
+        // Find any existing Products that match the desired ID
+        $args = array(
+            'posts_per_page'    => -1,
+            'post_type'         => 'wps-product',
+            'meta_key'          => '_wshop_product_id',
+            'meta_value'        => $id,
+            'post_status'       => 'publish,private,draft,future,pending'
```
Comment (johndigital) [inline wp-shopify.php:57]: I think you can just set this to 'any'

File: wp-shopify.php [hunk 7]
```diff
@@ -26,4 +26,132 @@ function pp() {
...
+                    '_wshop_product_id' => $id
+                )
+            );
+            wp_insert_post($args);
+
+            $output = 'Added new Product ' . $title . '.';
+
+        }
+
+        $output .= ' {ID:' . $id . '}';
```
Comment (johndigital) [inline wp-shopify.php]: You can either json_encode the output for this function, or just assume that it's a regular string. But I wouldn't do a mix of the two

File: js/wshop.refresh.js [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                product_id: data.id,
+                product_title: data.title,
+                auto_publish: jQuery('#auto_approve').is(':checked')
+            }
+        }).done(function(message){
+
+            jQuery('.refresh-message').append(`<li>(${wshopRefresh.totalProducts - wshopRefresh.products.length} / ${wshopRefresh.totalProducts}) ${message}</li>`);
+
+            // Strip out the product ID and save it to a list of IDs we've processed
+            var processedID = message.match(/\{ID:(\d+)\}/);
```
Comment (johndigital) [inline js/wshop.refresh.js]: See my note about json_encoding the output. This regex is definitely unnecessary

File: wshop-core.php [hunk 9]
```diff
@@ -1,6 +1,51 @@
...
+            'not_found_in_trash'    => 'No products found in Trash'
+        );
+
+        $args = array(
+            'labels'                => $labels,
+            'public'                => true,
+            'publicly_queryable'    => true,
+            'capability_type'       => 'page',
+            'menu_icon'             => 'dashicons-cart',
+            'hierarchical'          => true,
```
Comment (johndigital) [inline wshop-core.php]: Not actually sure that we need this now that I'm thinking about it. Let's talk about it

File: wshop-core.php [hunk 10]
```diff
@@ -34,6 +79,11 @@ function wshop_underscore_templates() {
         // find all custom templates (defined in the current theme)
         $user_template_paths = glob(get_stylesheet_directory() . '/wshop-templates/*.php');
 
+        // exit if no paths found
+        if( count($user_template_paths <= 0) ){
```
Comment (johndigital) [inline wshop-core.php]: Probably just wrap the following for each loop in this if statement rather than returning

File: wp-shopify.php (hunk 6 above)
Comment (SaFrMo) [reply wp-shopify.php:57]: @jrobson153 I didn't want to look in the trash - should it still be set to 'any'?

Comment (SaFrMo) [reply wp-shopify.php:57]: Didn't find any docs about "-trash" (or other 'except' indicators), so I'm leaving this as-is

**Pull request merged**